src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReadAheadWrappingStream.cs

```diff
@@ -0,0 +1,304 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+// 
+
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Runtime;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class ReadAheadWrappingStream : DelegatingStream
+    {
+        public const string ReadAheadWrappingStreamPropertyName = "ReadAheadWrappingStreamProperty";
+        private readonly int _readAheadBytes;
+        private readonly ArraySegment<byte>[] _buffers;
+        private Task<int> _previousCompletedReadTask;
+        private Task<int> _previousCompletedEnsureBufferedTask;
+        private bool _eof = false;
+
+        public ReadAheadWrappingStream(Stream innerStream, int readAheadBytes) : base(innerStream)
+        {
+            _readAheadBytes = readAheadBytes;
+            _buffers = new ArraySegment<byte>[2];
+            _buffers[0] = new ArraySegment<byte>(new byte[_readAheadBytes], 0, 0);
+            _buffers[1] = new ArraySegment<byte>(new byte[_readAheadBytes], 0, 0);
+            _previousCompletedReadTask = _previousCompletedEnsureBufferedTask = Task.FromResult(readAheadBytes);
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<int>(cancellationToken);
+            }
+
+            if (_buffers[0].Count == 0 && _buffers[1].Count == 0)
```

Comment:
The _previousCompletedReadTask is about avoiding creating a new Task object when completing synchronously if possible. For example, if a caller to ReadAsync asks for 1024 bytes, chances are it will ask for 1024 bytes again the next time. If this is the case, then we can avoid allocating a Task object by reusing the same one as was previously returned (if we can fulfill the request from the buffers). In this case, the buffers were empty so we need to call base.ReadAsync. There's no point trying to reuse the same Task as the base class has already provided the task. Plus it's only doable if the current operation has completed synchronously, otherwise you can't compare the result with the previous result, which won't be the case here. In this case, we just save the returned Task in case it's the same next time.